libsmtutil/Z3Loader.cpp

```diff
@@ -0,0 +1,77 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libsmtutil/Z3Loader.h>
+#include <libsmtutil/Exceptions.h>
+#include <z3.h>
+#include <z3_version.h>
+#include <vector>
+
+#ifndef _GNU_SOURCE
+#define _GNU_SOURCE
+#endif
+#include <dlfcn.h>
+
+using namespace std;
+
+Z3Loader const& Z3Loader::get()
+{
+	static Z3Loader z3;
+	return z3;
+}
+void* Z3Loader::loadSymbol(char const* _name) const
+{
+	using namespace solidity::smtutil;
+	smtAssert(m_handle, "Attempted to use dynamically loaded Z3, even though it is not available.");
+	void* sym = dlsym(m_handle, _name);
+	smtAssert(sym, string("Symbol \"") + _name + "\" not found in libz3.so");
+	return sym;
+}
+
+bool Z3Loader::available() const
+{
+	if (m_handle == nullptr)
+		return false;
+	unsigned major = 0;
+	unsigned minor = 0;
+	unsigned build = 0;
+	unsigned rev = 0;
+	Z3_get_version(&major, &minor, &build, &rev);
+	return major == 4 && minor >= 6;
+}
+
+Z3Loader::Z3Loader()
+{
+	vector<unsigned> versionParts{Z3_MAJOR_VERSION, Z3_MINOR_VERSION, Z3_BUILD_NUMBER, Z3_REVISION_NUMBER};
```

Comment:
Why don't we need to call `Z3_get_version` here?

Reply:
Where from? This looks for and loads ``libz3``, so at this point there *is* no ``Z3_get_version`` to call.
What it does is opening the the "best fit" of installed ``libz3.so.*`` most closely matching the build version. You can have multiple z3 libraries installed, imagine:

```
/usr/lib/libz3.so -> /usr/lib/libz3.so.4.8
/usr/lib/libz3.so.4.8 -> /usr/lib/libz3.so.4.8.9.0
/usr/lib/libz3.so.4.8.9.0
/usr/lib/libz3.so.4.8.7.0
```
is installed - then if we were built against 4.8.7.0, it'll load that one. If we were built against 4.8.8.0, then it'll load the 4.8 symlink. If it finds nothing, it'll try plain ``libz3.so`` - which is ok, since we still check ``Z3_get_version`` later (but we only can, once we loaded it).

Reply:
Ah ok, yes. Sorry, but I still don't get where `Z3_MAJOR_VERSION` and co are coming from.

Reply:
That's the build time version. ``z3_version.h`` defines those depending on which version it's coming from.

Reply:
Ok!